tests/pytests/functional/states/test_x509_v2.py

```diff
@@ -0,0 +1,2638 @@
+                "invalidityDate": "2022-11-18 13:37:00",
+            },
+        },
+        {
+            "serial_number": "013380",
+            "extensions": {
+                "CRLReason": "certificateHold",
+                "invalidityDate": "2022-11-18 13:37:00",
+            },
+        },
+        {
+            "serial_number": "013381",
+            "extensions": {
+                "CRLReason": "privilegeWithdrawn",
+                "invalidityDate": "2022-11-18 13:37:00",
+            },
+        },
+        {
+            "serial_number": "013382",
+            "extensions": {
+                "CRLReason": "aACompromise",
+                "invalidityDate": "2022-11-18 13:37:00",
+            },
+        },
+        {
+            "serial_number": "013383",
+            "extensions": {
+                "CRLReason": "removeFromCRL",
+            },
+        },
+    ]
+
+
+@pytest.fixture()
+def csr_args(tmp_path, rsa_privkey):
+    return {
+        "name": f"{tmp_path}/csr",
+        "private_key": rsa_privkey,
+        "CN": "success",
+    }
+
+
+@pytest.fixture()
+def csr_args_exts():
+    return {
+        "basicConstraints": "critical, CA:TRUE, pathlen:1",
+        "keyUsage": "critical, cRLSign, keyCertSign",
+        "extendedKeyUsage": "OCSPSigning",
+        "subjectKeyIdentifier": "hash",
+        "subjectAltName": "DNS:sub.salt.ca,email:sub@salt.ca",
+        "certificatePolicies": "1.2.4.5",
+        "policyConstraints": "requireExplicitPolicy:3",
+        "inhibitAnyPolicy": 2,
+        "nameConstraints": "permitted;IP:192.168.0.0/255.255.0.0,excluded;email:.com",
+        "noCheck": True,
+        "tlsfeature": "status_request",
+    }
+
+
+@pytest.fixture()
+def pk_args(tmp_path):
+    return {
+        "name": f"{tmp_path}/private_key",
+    }
+
+
+@pytest.fixture(params=["1234"])
+def existing_file(tmp_path, request):
+    text = request.param
+    if callable(text):
+        text = request.getfixturevalue(text.__name__)
+    test_file = tmp_path / "cert"
+    test_file.write_text(text)
+    yield test_file
+    # cleanup is done by tmp_path
+
+
+@pytest.fixture(params=[{}])
+def existing_cert(x509, cert_args, ca_key, rsa_privkey, request):
+    cert_args["private_key"] = rsa_privkey
+    cert_args.update(request.param)
+    ret = x509.certificate_managed(**cert_args)
+    subject = cert_args.get("subject", "CN={}".format(cert_args.get("CN")))
+    _assert_cert_created_basic(
+        ret,
+        cert_args["name"],
+        rsa_privkey,
+        ca_key,
+        encoding=cert_args.get("encoding", "pem"),
+        passphrase=cert_args.get("pkcs12_passphrase"),
+        subject=subject
+        if "signing_policy" not in cert_args
+        else "CN=from_signing_policy",
+    )
+    yield cert_args["name"]
+
+
+@pytest.fixture(params=[{}])
+def existing_cert_chain(x509, cert_args, ca_key, rsa_privkey, ca_cert, request):
+    cert_args["private_key"] = rsa_privkey
+    cert_args.update(request.param)
```

Comment:
It's not necessarily empty since it can be parametrized:

```python
@pytest.mark.usefixtures("existing_cert_chain")
@pytest.mark.parametrize(
    "existing_cert_chain",
    [
        {"encoding": "pem"},
        # [...]
    ],
    indirect=True,
)
def test_certificate_managed_existing_chain(x509, cert_args):
    ret = x509.certificate_managed(**cert_args)
    _assert_not_changed(ret)
```

Reply:
Ah, an indirect fixture. Pretty cool!